File: Sources/NotchAgent/Features/Providers/Claude/ClaudeTranscriptParser.swift
import Foundation

/// Aggregated stats for one Claude Code transcript file (`~/.claude/projects/<slug>/<session>.jsonl`).
/// Hour buckets keep memory flat and are all the session-block math needs.
struct ClaudeFileStat: Sendable {
    struct HourStat: Sendable {
        var tokens: TokenUsage = .zero
        var costUSD: Double = 0
        var messages: Int = 0
    }

    struct ModelStat: Sendable {
        var tokens: TokenUsage = .zero
        var costUSD: Double = 0
    }

    var hours: [Date: HourStat] = [:]
    var byModel: [String: ModelStat] = [:]
    var lastActivity: Date?
    var lastModel: String?
    var usageLineCount: Int = 0
    /// Dedup keys survive incremental re-parses of the same file.
    var seenKeys: Set<String> = []
}

enum ClaudeTranscriptParser {
    private struct Line: Decodable {
        struct Message: Decodable {
            struct Usage: Decodable {
                let inputTokens: Int?
                let outputTokens: Int?
                let cacheCreationInputTokens: Int?
                let cacheReadInputTokens: Int?
            }
            let id: String?
            let model: String?
            let usage: Usage?
        }
        let type: String?
        let timestamp: String?
        let requestId: String?
        let message: Message?
    }

    private static let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        return d
    }()

    private static let usageMarker = Data("\"usage\"".utf8)
    private static let assistantMarker = Data("\"assistant\"".utf8)

    /// Byte-level pre-filter: transcripts carry multi-megabyte lines (base64
    /// images, tool dumps) that can never be usage records. memchr-speed scans
    /// skip them before the expensive JSONDecoder ever runs — this is what
    /// keeps a 100 MB live transcript parseable every refresh.
    static func quickMatch(_ line: Data) -> Bool {
        line.range(of: usageMarker) != nil && line.range(of: assistantMarker) != nil
    }

    /// Tolerant: malformed lines are skipped, only assistant lines with usage count.
    /// Deduplicates by requestId/message.id (retries re-log usage).
    /// Incremental: pass `from`/`into` to parse only bytes appended since the
    /// last scan; returns the stat plus the consumed byte offset.
    static func parseFile(
        at url: URL,
        from offset: UInt64 = 0,
        into base: ClaudeFileStat? = nil
    ) throws -> (stat: ClaudeFileStat, consumed: UInt64) {
        var stat = base ?? ClaudeFileStat()

        let consumed = try JSONLReader.forEachLine(at: url, startingAt: offset) { data, _ in
            guard quickMatch(data),
                  let line = try? decoder.decode(Line.self, from: data),
                  line.type == "assistant",
                  let usage = line.message?.usage,
                  let timestampString = line.timestamp,
                  let timestamp = Timestamps.parseISO8601(timestampString)
            else { return }

            if let key = line.requestId ?? line.message?.id {
                guard stat.seenKeys.insert(key).inserted else { return }
            }

            let model = line.message?.model ?? "claude"
            let tokens = TokenUsage(
                input: usage.inputTokens ?? 0,
                output: usage.outputTokens ?? 0,
                cacheWrite: usage.cacheCreationInputTokens ?? 0,
                cacheRead: usage.cacheReadInputTokens ?? 0
            )

            let cost = PricingTable.costUSD(model: model, usage: tokens)
            let hour = timestamp.flooredToHour
            var bucket = stat.hours[hour] ?? .init()
            bucket.tokens += tokens
            bucket.costUSD += cost
            bucket.messages += 1
            stat.hours[hour] = bucket

            var modelStat = stat.byModel[model] ?? .init()
            modelStat.tokens += tokens
            modelStat.costUSD += cost
            stat.byModel[model] = modelStat

            if stat.lastActivity.map({ timestamp > $0 }) ?? true {
                stat.lastActivity = timestamp
                stat.lastModel = model
            }
            stat.usageLineCount += 1
        }
        return (stat, consumed)
    }
}

/// Incremental per-file cache: unchanged files hit the cache, grown files
/// parse only the appended bytes, truncated/rotated files re-parse fully.
/// Without this, the actively-written transcript (100+ MB during long
/// sessions) would be re-parsed whole on every refresh tick.
actor ClaudeScanCache {
    private struct Entry {
        var stamp: FileStamp
        var offset: UInt64
        var stat: ClaudeFileStat
    }

    private var entries: [String: Entry] = [:]

    func stat(for url: URL) throws -> ClaudeFileStat? {
        guard let stamp = FileStamp(url: url) else { return nil }
        let key = url.path

        if let entry = entries[key] {
            if entry.stamp == stamp {
                return entry.stat
            }
            if UInt64(stamp.size) >= entry.offset {
                let (stat, consumed) = try ClaudeTranscriptParser.parseFile(
                    at: url, from: entry.offset, into: entry.stat
                )
                entries[key] = Entry(stamp: stamp, offset: consumed, stat: stat)
                return stat
            }
            // File shrank — rotated or rewritten. Fall through to a full parse.
        }
        let (stat, consumed) = try ClaudeTranscriptParser.parseFile(at: url)
        entries[key] = Entry(stamp: stamp, offset: consumed, stat: stat)
        return stat
    }

    /// Drops entries for files no longer in the scan set (deleted or aged out
    /// of the lookback), bounding memory for long-running sessions.
    func prune(keeping paths: Set<String>) {
        entries = entries.filter { paths.contains($0.key) }
    }
}

/// Claude Plans meter usage in rolling ~5h session blocks. This reproduces the
/// commonly used heuristic (same as ccusage): a block starts at the floored hour
/// of the first activity after the previous block ended, and lasts 5 hours.
enum SessionBlocks {
    static let blockLength: TimeInterval = 5 * 3600

    /// Returns the active block containing `now`, or nil when idle.
    static func currentBlock(activityHours: [Date], now: Date) -> (start: Date, end: Date)? {
        var blockStart: Date?
        var blockEnd = Date.distantPast

        for hour in activityHours.sorted() {
            if blockStart == nil || hour >= blockEnd {
                blockStart = hour
                blockEnd = hour.addingTimeInterval(blockLength)
            }
        }
        guard let start = blockStart, now >= start, now < blockEnd else { return nil }
        return (start, blockEnd)
    }
}

File: windows/NotchAgent.Windows/Providers/Claude/ClaudeTranscriptParser.cs
using System.Text;
using System.Text.Json;
using NotchAgent.Windows.Models;
using NotchAgent.Windows.Providers.Shared;

namespace NotchAgent.Windows.Providers.Claude;

public sealed class ClaudeFileStat
{
    public sealed class HourStat
    {
        public TokenUsage Tokens;
        public double CostUsd;
        public int Messages;
    }

    public sealed class ModelStat
    {
        public TokenUsage Tokens;
        public double CostUsd;
    }

    public Dictionary<DateTimeOffset, HourStat> Hours { get; } = new();
    public Dictionary<string, ModelStat> ByModel { get; } = new();
    public DateTimeOffset? LastActivity { get; set; }
    public string? LastModel { get; set; }
    public int UsageLineCount { get; set; }
    /// Dedup keys survive incremental re-parses of the same file.
    public HashSet<string> SeenKeys { get; } = new();
}

public static class ClaudeTranscriptParser
{
    private static readonly byte[] UsageMarker = Encoding.UTF8.GetBytes("\"usage\"");
    private static readonly byte[] AssistantMarker = Encoding.UTF8.GetBytes("\"assistant\"");

    /// Byte-level pre-filter: transcripts carry multi-megabyte lines (base64
    /// images, tool dumps) that can never be usage records. This skips them
    /// before the expensive JSON parse ever runs.
    public static bool QuickMatch(ReadOnlySpan<byte> line) =>
        line.IndexOf(UsageMarker) >= 0 && line.IndexOf(AssistantMarker) >= 0;

    public static DateTimeOffset FlooredToHour(DateTimeOffset date)
    {
        var epoch = date.ToUnixTimeSeconds();
        var flooredHours = (epoch / 3600) * 3600;
        return DateTimeOffset.FromUnixTimeSeconds(flooredHours);
    }

    /// Tolerant: malformed lines are skipped, only assistant lines with usage count.
    /// Deduplicates by requestId/message.id. Incremental: pass `from`/`into` to
    /// parse only bytes appended since the last scan.
    public static (ClaudeFileStat Stat, long Consumed) ParseFile(string path, long offset = 0, ClaudeFileStat? baseStat = null)
    {
        var stat = baseStat ?? new ClaudeFileStat();

        long consumed = JsonlReader.ForEachLine(path, offset, lineBytes =>
        {
            if (!QuickMatch(lineBytes.Span)) return;

            JsonDocument doc;
            try { doc = JsonDocument.Parse(lineBytes); }
            catch (JsonException) { return; }
            using (doc)
            {
                var root = doc.RootElement;
                if (!root.TryGetProperty("type", out var typeProp) || typeProp.GetString() != "assistant") return;
                if (!root.TryGetProperty("message", out var message)) return;
                if (!message.TryGetProperty("usage", out var usage)) return;
                if (!root.TryGetProperty("timestamp", out var tsProp)) return;
                if (!DateTimeOffset.TryParse(tsProp.GetString(), out var timestamp)) return;

                string? key = root.TryGetProperty("requestId", out var rid) ? rid.GetString()
                    : message.TryGetProperty("id", out var mid) ? mid.GetString() : null;
                if (key is not null && !stat.SeenKeys.Add(key)) return;

                string model = message.TryGetProperty("model", out var m) ? (m.GetString() ?? "claude") : "claude";
                var tokens = new TokenUsage
                {
                    Input = GetLong(usage, "input_tokens"),
                    Output = GetLong(usage, "output_tokens"),
                    CacheWrite = GetLong(usage, "cache_creation_input_tokens"),
                    CacheRead = GetLong(usage, "cache_read_input_tokens"),
                };
                var cost = PricingTable.CostUsd(model, tokens);

                var hour = FlooredToHour(timestamp);
                if (!stat.Hours.TryGetValue(hour, out var bucket))
                {
                    bucket = new ClaudeFileStat.HourStat();
                    stat.Hours[hour] = bucket;
                }
                bucket.Tokens += tokens;
                bucket.CostUsd += cost;
                bucket.Messages += 1;

                if (!stat.ByModel.TryGetValue(model, out var modelStat))
                {
                    modelStat = new ClaudeFileStat.ModelStat();
                    stat.ByModel[model] = modelStat;
                }
                modelStat.Tokens += tokens;
                modelStat.CostUsd += cost;

                if (stat.LastActivity is null || timestamp > stat.LastActivity)
                {
                    stat.LastActivity = timestamp;
                    stat.LastModel = model;
                }
                stat.UsageLineCount += 1;
            }
        });

        return (stat, consumed);
    }

    private static long GetLong(JsonElement obj, string prop) =>
        obj.TryGetProperty(prop, out var v) && v.TryGetInt64(out var n) ? n : 0;
}

/// Incremental per-file cache: unchanged files hit the cache, grown files
/// parse only the appended bytes, truncated/rotated files re-parse fully.
public sealed class ClaudeScanCache
{
    private sealed record Entry(FileStamp Stamp, long Offset, ClaudeFileStat Stat);

    private readonly Dictionary<string, Entry> _entries = new();
    private readonly object _lock = new();

    public ClaudeFileStat? Stat(string path)
    {
        var stamp = FileStamp.Of(path);
        if (stamp is not { } s) return null;

        lock (_lock)
        {
            if (_entries.TryGetValue(path, out var entry))
            {
                if (entry.Stamp.Equals(s)) return entry.Stat;
                if (s.Size >= entry.Offset)
                {
                    var (stat, consumed) = ClaudeTranscriptParser.ParseFile(path, entry.Offset, entry.Stat);
                    _entries[path] = new Entry(s, consumed, stat);
                    return stat;
                }
                // File shrank — rotated or rewritten. Fall through to a full parse.
            }
            var (fullStat, fullConsumed) = ClaudeTranscriptParser.ParseFile(path);
            _entries[path] = new Entry(s, fullConsumed, fullStat);
            return fullStat;
        }
    }

    /// Drops entries for files no longer in the scan set.
    public void Prune(HashSet<string> keep)
    {
        lock (_lock)
        {
            foreach (var key in _entries.Keys.Where(k => !keep.Contains(k)).ToList())
            {
                _entries.Remove(key);
            }
        }
    }
}
